codec-redis/src/main/java/io/netty/handler/codec/redis/RedisMessageFactory.java

```diff
@@ -0,0 +1,144 @@
+        return msg != null ? msg : new StringRedisMessage(RedisMessageType.SIMPLE_STRING, content);
+    }
+
+    public static RedisMessage createSimpleString(ByteBuf content) {
+        RedisMessage msg = findPredefinedString(PREDEFINED_SIMPLE_STRINGS, content);
+        return msg != null ? msg : new StringRedisMessage(RedisMessageType.SIMPLE_STRING,
+                                                          content.toString(CharsetUtil.UTF_8));
+    }
+
+    public static RedisMessage createError(String content) {
+        RedisMessage msg = findPredefinedString(PREDEFINED_ERRORS, content);
+        return msg != null ? msg : new StringRedisMessage(RedisMessageType.ERROR, content);
+    }
+
+    public static RedisMessage createError(ByteBuf content) {
+        RedisMessage msg = findPredefinedString(PREDEFINED_ERRORS, content);
+        return msg != null ? msg : new StringRedisMessage(RedisMessageType.ERROR, content.toString(CharsetUtil.UTF_8));
+    }
+
+    public static RedisMessage createInteger(long value) {
+        return new IntegerRedisMessage(value);
+    }
+
+    public static RedisMessage nullBulkString() {
+        return NULL_BULK_STRING;
+    }
+
+    public static RedisMessage emptyBulkString() {
+        return EMPTY_BULK_STRING;
+    }
+
+    public static RedisMessage createBulkString(ByteBuf byteBuf) {
+        if (byteBuf == null) {
+            return NULL_BULK_STRING;
+        } else if (!byteBuf.isReadable()) {
+            return EMPTY_BULK_STRING;
+        }
+        return new RefCountedBulkStringRedisMessage(byteBuf);
+    }
+
+    public static RedisMessage createArrayHeader(long length) {
+        return new ArrayHeaderRedisMessage(length);
+    }
+
+    public static RedisMessage nullArray() {
+        return NULL_ARRAY;
+    }
+
+    public static RedisMessage emptyArray() {
+        return EMPTY_ARRAY;
+    }
+
+    public static RedisMessage createArray(List<RedisMessage> children) {
+        if (children == null) {
+            return NULL_ARRAY;
+        } else if (children.isEmpty()) {
+            return EMPTY_ARRAY;
+        } else {
+            return new RefCountedArrayRedisMessage(children);
+        }
+    }
+
+    private static RedisMessage findPredefinedString(PredefinedString[] predefinedStrings, ByteBuf byteBuf) {
+        for (PredefinedString predefined : predefinedStrings) {
+            if (ByteBufUtil.equals(predefined.byteBuf, byteBuf)) {
+                return predefined.message;
+            }
+        }
+        return null;
+    }
+
+    private static RedisMessage findPredefinedString(PredefinedString[] predefinedStrings, String content) {
+        for (PredefinedString predefined : predefinedStrings) {
+            if (predefined.content.equals(content)) {
+                return predefined.message;
+            }
+        }
+        return null;
+    }
+
+    private static class PredefinedString {
```

Comment:
We could make this type and fields package-protected so the JVM doesn't create synthetic accessors.